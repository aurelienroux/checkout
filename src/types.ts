export interface product {
  id: string
  name: string
  image: string
  variants: variant[]
  description: string
  defaultVariantId: string
}

export interface variant {
  id: string
  name: string
  price: number
  attributes: {
    name: string
    value: string
  }[]
}

export interface updateAttribute {
  attribute: string
  value: string
}

export interface productToAdd {
  description: string
  image: string
  name: string
  price: number
  productId: string
  variantId: string
}

export interface productInCart {
  description: string
  price: number
  productId: string
  qty: number
  variantId: string
}

export enum attributeEnum {
  SIZE = 'Size',
  FABRIC = 'Fabric'
}

export enum cartActionEnum {
  ADD = 'add',
  REMOVE = 'remove'
}
