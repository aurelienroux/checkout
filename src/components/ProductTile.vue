<template>
  <div>
    {{ productInfo.name }}
    <div>selectedVariantId: {{ selectedVariantId }}</div>
    <div>Price : ${{ currentPrice }}</div>
    <!-- <div>laSoluce: {{ possibleVariants }}</div> -->
    <AttributeSelector v-for="(variant, index) in possibleVariants" :key="index" :variant="variant" />
    <hr />
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import AttributeSelector from './AttributeSelector'

export default Vue.extend({
  name: 'ProductTile',
  props: ['productInfo'],
  components: {
    AttributeSelector
  },
  computed: {
    selectedVariantId() {
      return this.productInfo.defaultVariantId
    },
    currentPrice() {
      const selectedVariant = this.productInfo.variants.filter(variant => {
        return variant.id === this.selectedVariantId
      })
      return selectedVariant[0].price
    },
    possibleVariants() {
      const result = []

      this.productInfo.variants.forEach(({ attributes }) => {
        attributes.forEach(attribute => {
          const { name, value } = attribute
          const index = result.findIndex(el => el['name'] === name)

          if (index !== -1) {
            if (result[index].values.includes(value)) return

            result[index].values.push(value)
          } else {
            result.push({
              name: name,
              values: [value]
            })
          }
        })
      })

      return result
    }
  }
})
</script>
