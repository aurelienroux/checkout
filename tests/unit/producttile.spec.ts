import { shallowMount } from '@vue/test-utils'

import ProductTile from '@/components/ProductTile.vue'
import { attributeEnum, product } from '@/types'

const product: product = {
  id: '1',
  name: 'product name',
  image: 'http://image.example',
  variants: [
    {
      id: '1',
      name: '36 / cotton',
      price: 99,
      attributes: [
        {
          name: attributeEnum.SIZE,
          value: '36'
        },
        {
          name: attributeEnum.FABRIC,
          value: 'cotton'
        }
      ]
    }
  ],
  description: 'description',
  defaultVariantId: '1'
}

const wrapper = shallowMount(ProductTile, {
  propsData: {
    productInfo: product
  },
  computed: {
    selectedVariant: () => product.variants[0],
    sizeOptions: () => ['36', '38'],
    fabricOptions: () => ['cotton', 'leather']
  }
})

describe('ProductTile component', () => {
  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.tile__image').exists()).toBe(true)
    expect(wrapper.find('.tile__content').exists()).toBe(true)
    expect(wrapper.find('img').exists()).toBe(true)
    expect(wrapper.find('img').attributes('src')).toEqual(product.image)
    expect(wrapper.find('.button--purple').exists()).toBe(true)
    expect(wrapper.find('.button--purple').text()).toEqual('add to cart')
  })

  it('renders with correct header', () => {
    expect(wrapper.find('header').exists()).toBe(true)
    expect(wrapper.find('header .price').exists()).toBe(true)
    expect(wrapper.find('header h2').text()).toEqual(product.name)
    expect(wrapper.find('header .price').text()).toEqual(`$${product.variants[0].price}`)
  })

  it('renders with description', () => {
    expect(wrapper.find('p').text()).toEqual(product.description)
  })

  it('renders with correct attributes', () => {
    const selector = 'attributeselector-stub'

    expect(wrapper.findAll(selector).length).toBe(2)
    expect(wrapper.findAll(selector).at(0).attributes('attribute')).toEqual('Size')
    expect(wrapper.findAll(selector).at(0).attributes('attroptions')).toEqual('36,38')
    expect(wrapper.findAll(selector).at(0).attributes('selectedvalue')).toEqual('36')
    expect(wrapper.findAll(selector).at(1).attributes('attribute')).toEqual('Fabric')
    expect(wrapper.findAll(selector).at(1).attributes('attroptions')).toEqual('cotton,leather')
    expect(wrapper.findAll(selector).at(1).attributes('selectedvalue')).toEqual('cotton')
  })
})
