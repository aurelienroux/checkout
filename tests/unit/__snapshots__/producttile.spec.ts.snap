// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTile component renders without crashing 1`] = `
<div class="tile container">
  <div class="tile__image"><img src="http://image.example" alt="product name"></div>
  <div class="tile__content">
    <header>
      <h2>product name</h2>
      <div class="price">$99</div>
    </header>
    <p>description</p>
    <div class="attributes">
      <attributeselector-stub attribute="Size" attroptions="36,38" selectedvalue="36"></attributeselector-stub>
      <attributeselector-stub attribute="Fabric" attroptions="cotton,leather" selectedvalue="cotton"></attributeselector-stub>
    </div> <button class="button button--purple">add to cart</button>
  </div>
</div>
`;
